### ISSUE TITLE
Binding to the first free port in a given range is trappy

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#54354]
Today `transport.port` and `http.port` accept a range of ports. At startup, Elasticsearch binds to the first available port in the range. This is useful for testing because it lets you stand up a number of nodes on a single host without any fancy configuration.

In a production environment with multi-node hosts this behaviour can be trappy, because it means you have no idea which specific node you're talking to at any given port. The choice of node is important because you should generally avoid sending client requests to dedicated master nodes and generally prefer sending indexing or search requests to specific coordinating nodes. It'll still work, but will not give you the performance you expect.

This also has an impact on discovery. The default discovery port is 9300 and as long as there's at least one master-eligible node running on port 9300 a naively-configured discovery process will work. However there is a chance that no master-eligible nodes manage to bind to the default port and in this case the cluster will not form, and it can be tricky to work out why.

For this reason I think we should stop accepting port ranges in production environments. Instead, we should either bind to a specific port or die trying.

**Comment by DaveCTurner** [comment]
We discussed this today and broadly agreed. We may want to get rid of port ranges altogether, they make life complicated and we are not convinced of their value to the OOTB experience.

I will speak to some training folk since I know we use this feature in today's training courses.

**Comment by DaveCTurner** [comment]
I confirmed with the training folk that this feature is no longer needed in today's training courses.

**Comment by kingherc** [comment]
Would this be considered a breaking change?

**Comment by DaveCTurner** [comment]
Yes, we would have to do the usual deprecation dance: in 8.x we should emit a deprecation warning if using a port range (including not specifying a port) to raise awareness of the change in behaviour. Then in 9.0 we can drop support for port ranges.

**Comment by kingherc** [comment]
If it also needs to pass the deprecation cycle, sounds more than tech debt, since it seems user-facing and will need some time to handle. Maybe we can add the >enhancement label as well, will do so (similar to https://github.com/elastic/elasticsearch/issues/84563 ), feel free to alter if it's a bad move.
